First try at anonymization

psorianom opened the pull request:
Kinda finished the anonymization portion of the code (anon and fix commands).
This gives something like :

| replace                 | with        |
|-------------------------|-------------|
| Salomon,Hirsch,Vareille | XXX,XXX,XXX |
| Marcovici               | XXX         |
| Joseph,Maryse           | XXX,XXX     |
| MARIN                   | XXX         |
| Paul                    | XXX         |
| Gilles                  | XXX         |
| Fran                    | XXX         |
| Yvette,Alain,Yvette     | XXX,XXX,XXX |
| Erre                    | XXX         |
| Claude,Claude           | XXX,XXX     |

abulte commented on cada/commands.py:
```diff
@@ -205,15 +206,16 @@ def anon():
     '''Check for candidates to anonymization'''
     header(anon.__doc__)
     filename = 'urls_to_check.csv'
+    name_regex = r'(M\.|Mme\.|Monsieur|Madame|Docteur|Mademoiselle)\s+[^X\s\.]{3}'
```
Add `Mr` I guess. It’s not proper French but a lot of people make the mistake.

taniki replied:
"Mlle" can be added to

abulte replied:
« Maître » and whatever is its abbreviation(s)

abulte replied:
`Mme` w/o final point.

abulte replied:
`Dr` for Docteur

taniki replied:
http://cada.data.gouv.fr/20120785/ peut être ajouter l'anonymisation des entreprises

noirbizarre commented:
Be carefull, the `fix` command expect a name by line so `Salomon,Hirsch,Vareille` isn't valid.
The `fix` command needs to have uniques match so replacement chars are the same for a given name (ie. 'Salomon' needs to be always be replaced by `XXX` and `Hirsch` by `YYY`)

psorianom commented on cada/commands.py:
```diff
@@ -223,9 +225,21 @@ def anon():
+                matches = re.findall(match_regex, possible_location)
+                if not matches:
+                    continue
+                good_matches = [m[1] for m in matches if len(m) > 1 and len(m[1]) > 3 and m[1][0].isupper()]
```
With this type of queries on MongoDB, can you recover within the `candidate` variable  the actual match of these regexes ? If not, I still need to re-read both (advice and subject) looking for the actual match.

psorianom replied:
If I understood correctly, we want to get those things that are not XXX (`name_regex1`) and then filter them with `name_regex2`. I tried doing this quickly, but it does not seem to work:

```
    candidates = Advice.objects(__raw__={

        '$and': [{
                "$or": [{
                    'subject': {
                        '$regex': name_regex1,
                        '$options': 'imx',
                    }
                },
                    {
                    'content': {
                        '$regex': name_regex1,
                        '$options': 'imx',
                    }}
                ]},

                {
                '$or': [{
                    'subject': {
                        '$regex': match_regex,
                        '$options': 'mx',
                    }
                },
                    {
                    'content': {
                        '$regex': match_regex,
                        '$options': 'mx',
                        }
                    }
                ]}
            ]})
```

I think that a single regex like this `match_regex = r'(%s)\s+([A-WYZ][^X\.\-][\w]+)' % PREFIXES` would do the job. The dot and dash to avoid _P.a_ or _Z-M_ (already anonymized text).

What do you think ?

taniki replied:
that seems ok to me

psorianom commented:
> Be carefull, the `fix` command expect a name by line so `Salomon,Hirsch,Vareille` isn't valid.
> The `fix` command needs to have uniques match so replacement chars are the same for a given name (ie. 'Salomon' needs to be always be replaced by `XXX` and `Hirsch` by `YYY`)

I did it that way to comply with the code on `fix()`

```
sources = [s.strip() for s in sources.split(',') if s.strip()]
dests = [d.strip() for d in dests.split(',') if d.strip()]

```

psorianom commented:
Salut! 
I am using now a pipeline to perform an aggregation while querying mongodb. There are is a new feature (or issues :/ ) which may or may not be solved somehow which I am not aware of: the aggregation query result is an iterable of dictionaries. That is, we no longer receive an iterable of classes of type `Advice` but a dictionary with the elements of a regular Advice instance but in a key-value shape.

This causes some of the code that follows to fail as elements from Advice are accessed as a class (`advice.id`) instead of as a dict (`advice['_id']`). 
I managed those errors but maybe you will not like how it is done (the elements of the objects that are an instance of `Advice` can be accessed via itemkey `Advice['_id']`.

I kinda do not know how to show just my last commit. Some cherry-picking must be done I guess :/

abulte commented:
@psorianom si tu es prêt à repasser sur ce code et voir si tout est OK pour toi, je suis prêt à faire une _review_ ;-)
